File: ee25btech11047/Matgeo/problem20/codes/12.164.py
import numpy as np

# Define matrix A
A = np.array([
    [1, 1, 1],
    [0, 1, 1],
    [0, 0, 1]
], dtype=float)

# Step 1: Eigenvalues
eigenvalues, _ = np.linalg.eig(A)
print("Eigenvalues:")
for i, val in enumerate(eigenvalues, start=1):
    print(f"λ{i} = {val:.2f}")

# Step 2: For eigenvalue = 1, compute (A - I)
I = np.eye(3)
AI = A - I

print("\nMatrix (A - I):")
print(AI)

# Step 3: Find null space to get eigenvectors
# (A - I)x = 0 → eigenvectors for λ = 1
u, s, vh = np.linalg.svd(AI)
tol = 1e-10
null_mask = (s <= tol)
null_space = np.compress(null_mask, vh, axis=0)

print("\nEigenvectors corresponding to λ = 1:")
if null_space.size == 0:
    print("No eigenvectors found.")
else:
    print(null_space.T)

# Step 4: Number of linearly independent eigenvectors
num_independent = null_space.shape[0]
print(f"\nMaximum number of linearly independent eigenvectors = {num_independent}")

File: ee25btech11047/Matgeo/problem20/codes/12.164.c
#include <stdio.h>
#include <math.h>

int main() {
    // Matrix A
    double A[3][3] = {
        {1, 1, 1},
        {0, 1, 1},
        {0, 0, 1}
    };

    // Step 1: Since A is upper triangular, eigenvalues are diagonal elements
    double lambda[3];
    for (int i = 0; i < 3; i++) {
        lambda[i] = A[i][i];
    }

    printf("Eigenvalues:\n");
    for (int i = 0; i < 3; i++) {
        printf("λ%d = %.2f\n", i+1, lambda[i]);
    }

    // Step 2: All eigenvalues are equal (1), find eigenvectors for λ = 1
    // (A - I)
    double AI[3][3];
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            AI[i][j] = A[i][j];
            if (i == j) AI[i][j] -= 1.0;
        }
    }

    printf("\nMatrix (A - I):\n");
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            printf("%6.2f ", AI[i][j]);
        }
        printf("\n");
    }

    // From manual solving:
    // x3 = 0, x2 = 0, x1 = free → 1 eigenvector
    printf("\nEigenvector corresponding to λ = 1:\n");
    printf("v1 = [1 0 0]^T\n");

    printf("\nMaximum number of linearly independent eigenvectors = 1\n");
    return 0;
}

